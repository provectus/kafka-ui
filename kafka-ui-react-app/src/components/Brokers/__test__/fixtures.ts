import { BrokersLogdirs } from 'generated-sources';

export const brokersPayload = [
  { id: 1, host: 'b-1.test.kafka.amazonaws.com', port: 9092 },
  { id: 2, host: 'b-2.test.kafka.amazonaws.com', port: 9092 },
];

export const clusterStatsPayload = {
  brokerCount: 2,
  activeControllers: 1,
  onlinePartitionCount: 138,
  offlinePartitionCount: 0,
  inSyncReplicasCount: 239,
  outOfSyncReplicasCount: 0,
  underReplicatedPartitionCount: 0,
  diskUsage: [
    { brokerId: 0, segmentSize: 334567, segmentCount: 245 },
    { brokerId: 1, segmentSize: 12345678, segmentCount: 121 },
  ],
  version: '2.2.1',
};

const partition = {
  broker: 2,
  offsetLag: 0,
  partition: 2,
  size: 0,
};
const topics = {
  name: '_confluent-ksql-devquery_CTAS_NUMBER_OF_TESTS_59-Aggregate-Aggregate-Materialize-changelog',
  partitions: [partition],
};

export const brokerLogDirsPayload: BrokersLogdirs[] = [
  {
    error: 'NONE',
    name: '/opt/kafka/data-0/logs',
    topics: [
      {
        ...topics,
        partitions: [partition, partition, partition],
      },
      topics,
      {
        ...topics,
        partitions: [],
      },
    ],
  },
];
