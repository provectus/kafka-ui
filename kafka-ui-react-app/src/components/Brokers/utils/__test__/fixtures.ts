import { BrokerMetrics } from 'generated-sources';

export const brokerMetricsPayload: BrokerMetrics = {
  segmentSize: 23,
  segmentCount: 23,
  metrics: [
    {
      name: 'TotalFetchRequestsPerSec',
      labels: {
        canonicalName:
          'kafka.server:name=TotalFetchRequestsPerSec,topic=_connect_status,type=BrokerTopicMetrics',
      },
      value: 10,
    },
    {
      name: 'ZooKeeperRequestLatencyMs',
      value: 11,
    },
    {
      name: 'RequestHandlerAvgIdlePercent',
    },
  ],
};
export const transformedBrokerMetricsPayload =
  '{"segmentSize":23,"segmentCount":23,"metrics":[{"name":"TotalFetchRequestsPerSec","labels":{"canonicalName":"kafka.server:name=TotalFetchRequestsPerSec,topic=_connect_status,type=BrokerTopicMetrics"},"value":10},{"name":"ZooKeeperRequestLatencyMs","value":11},{"name":"RequestHandlerAvgIdlePercent"}]}';
