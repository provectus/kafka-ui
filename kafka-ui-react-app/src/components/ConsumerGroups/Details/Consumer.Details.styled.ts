import styled from 'styled-components';

export const Drop = styled.span`
  color: ${({ theme }) => theme.dropdown.color};
`;
