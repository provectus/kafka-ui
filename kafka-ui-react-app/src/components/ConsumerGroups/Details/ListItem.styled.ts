import styled from 'styled-components';

export const FlexWrapper = styled.div`
  display: flex;
  align-items: center;
  gap: 8px;
`;
