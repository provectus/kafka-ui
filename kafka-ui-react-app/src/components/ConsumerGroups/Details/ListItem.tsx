import React from 'react';
import { ConsumerGroupTopicPartition } from 'generated-sources';
import { Link } from 'react-router-dom';
import { ClusterName } from 'redux/interfaces/cluster';
import { clusterTopicPath } from 'lib/paths';
import MessageToggleIcon from 'components/common/Icons/MessageToggleIcon';
import IconButtonWrapper from 'components/common/Icons/IconButtonWrapper';
import { TableKeyLink } from 'components/common/table/Table/TableKeyLink.styled';

import TopicContents from './TopicContents/TopicContents';
import { FlexWrapper } from './ListItem.styled';

interface Props {
  clusterName: ClusterName;
  name: string;
  consumers: ConsumerGroupTopicPartition[];
}

const ListItem: React.FC<Props> = ({ clusterName, name, consumers }) => {
  const [isOpen, setIsOpen] = React.useState(false);

  const getTotalMessagesBehind = () => {
    let count = 0;
    consumers.forEach((consumer) => {
      count += consumer?.messagesBehind || 0;
    });
    return count;
  };

  return (
    <>
      <tr>
        <td>
          <FlexWrapper>
            <IconButtonWrapper onClick={() => setIsOpen(!isOpen)} aria-hidden>
              <MessageToggleIcon isOpen={isOpen} />
            </IconButtonWrapper>
            <TableKeyLink>
              <Link to={clusterTopicPath(clusterName, name)}>{name}</Link>
            </TableKeyLink>
          </FlexWrapper>
        </td>
        <td>{getTotalMessagesBehind()}</td>
      </tr>
      {isOpen && <TopicContents consumers={consumers} />}
    </>
  );
};

export default ListItem;
