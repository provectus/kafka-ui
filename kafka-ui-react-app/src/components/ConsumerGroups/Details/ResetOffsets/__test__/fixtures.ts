export const expectedOutputs = {
  EARLIEST: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: undefined,
        partition: 0,
      },
    ],
    resetType: 'EARLIEST',
    topic: '__amazon_msk_canary',
  },
  LATEST: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: undefined,
        partition: 0,
      },
    ],
    resetType: 'LATEST',
    topic: '__amazon_msk_canary',
  },
  OFFSET: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: '10',
        partition: 0,
      },
    ],
    resetType: 'OFFSET',
    topic: '__amazon_msk_canary',
  },
};
