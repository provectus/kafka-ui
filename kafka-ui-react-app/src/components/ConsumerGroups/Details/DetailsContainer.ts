import { connect } from 'react-redux';
import Details from './Details';
import {ClusterName, RootState} from 'redux/interfaces';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { getIsConsumerGroupDetailsFetched, getConsumerGroupByID } from 'redux/reducers/consumerGroups/selectors';
import { ConsumerGroupID } from 'redux/interfaces/consumerGroup';
import { fetchConsumerGroupDetails } from 'redux/actions/thunks';

interface RouteProps {
  clusterName: ClusterName;
  consumerGroupID: ConsumerGroupID;
}

interface OwnProps extends RouteComponentProps<RouteProps> { }

const mapStateToProps = (state: RootState, { match: { params: { consumerGroupID, clusterName } } }: OwnProps) => ({
  clusterName,
  consumerGroupID,
  isFetched: getIsConsumerGroupDetailsFetched(state),
  ...getConsumerGroupByID(state, consumerGroupID)
});

const mapDispatchToProps = {
  fetchConsumerGroupDetails: (clusterName: ClusterName, consumerGroupID: ConsumerGroupID) => fetchConsumerGroupDetails(clusterName, consumerGroupID),
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(Details)
);
