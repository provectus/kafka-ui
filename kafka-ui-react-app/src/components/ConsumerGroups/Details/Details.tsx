import React from 'react';
import { useNavigate, useSearchParams } from 'react-router-dom';
import useAppParams from 'lib/hooks/useAppParams';
import {
  clusterConsumerGroupResetRelativePath,
  clusterConsumerGroupsPath,
  ClusterGroupParam,
} from 'lib/paths';
import Search from 'components/common/Search/Search';
import ClusterContext from 'components/contexts/ClusterContext';
import PageHeading from 'components/common/PageHeading/PageHeading';
import * as Metrics from 'components/common/Metrics';
import { Tag } from 'components/common/Tag/Tag.styled';
import groupBy from 'lodash/groupBy';
import { Table } from 'components/common/table/Table/Table.styled';
import getTagColor from 'components/common/Tag/getTagColor';
import { Dropdown } from 'components/common/Dropdown';
import { ControlPanelWrapper } from 'components/common/ControlPanel/ControlPanel.styled';
import { Action, ResourceType } from 'generated-sources';
import { ActionDropdownItem } from 'components/common/ActionComponent';
import TableHeaderCell from 'components/common/table/TableHeaderCell/TableHeaderCell';
import {
  useConsumerGroupDetails,
  useDeleteConsumerGroupMutation,
} from 'lib/hooks/api/consumers';

import ListItem from './ListItem';

const Details: React.FC = () => {
  const navigate = useNavigate();
  const [searchParams] = useSearchParams();
  const searchValue = searchParams.get('q') || '';
  const { isReadOnly } = React.useContext(ClusterContext);
  const routeParams = useAppParams<ClusterGroupParam>();
  const { clusterName, consumerGroupID } = routeParams;

  const consumerGroup = useConsumerGroupDetails(routeParams);
  const deleteConsumerGroup = useDeleteConsumerGroupMutation(routeParams);

  const onDelete = async () => {
    await deleteConsumerGroup.mutateAsync();
    navigate('../');
  };

  const onResetOffsets = () => {
    navigate(clusterConsumerGroupResetRelativePath);
  };

  const partitionsByTopic = groupBy(consumerGroup.data?.partitions, 'topic');
  const filteredPartitionsByTopic = Object.keys(partitionsByTopic).filter(
    (el) => el.includes(searchValue)
  );
  const currentPartitionsByTopic = searchValue.length
    ? filteredPartitionsByTopic
    : Object.keys(partitionsByTopic);

  return (
    <div>
      <div>
        <PageHeading
          text={consumerGroupID}
          backTo={clusterConsumerGroupsPath(clusterName)}
          backText="Consumers"
        >
          {!isReadOnly && (
            <Dropdown>
              <ActionDropdownItem
                onClick={onResetOffsets}
                permission={{
                  resource: ResourceType.CONSUMER,
                  action: Action.RESET_OFFSETS,
                  value: consumerGroupID,
                }}
              >
                Reset offset
              </ActionDropdownItem>
              <ActionDropdownItem
                confirm="Are you sure you want to delete this consumer group?"
                onClick={onDelete}
                danger
                permission={{
                  resource: ResourceType.CONSUMER,
                  action: Action.DELETE,
                  value: consumerGroupID,
                }}
              >
                Delete consumer group
              </ActionDropdownItem>
            </Dropdown>
          )}
        </PageHeading>
      </div>
      <Metrics.Wrapper>
        <Metrics.Section>
          <Metrics.Indicator label="State">
            <Tag color={getTagColor(consumerGroup.data?.state)}>
              {consumerGroup.data?.state}
            </Tag>
          </Metrics.Indicator>
          <Metrics.Indicator label="Members">
            {consumerGroup.data?.members}
          </Metrics.Indicator>
          <Metrics.Indicator label="Assigned Topics">
            {consumerGroup.data?.topics}
          </Metrics.Indicator>
          <Metrics.Indicator label="Assigned Partitions">
            {consumerGroup.data?.partitions?.length}
          </Metrics.Indicator>
          <Metrics.Indicator label="Coordinator ID">
            {consumerGroup.data?.coordinator?.id}
          </Metrics.Indicator>
          <Metrics.Indicator label="Total lag">
            {consumerGroup.data?.messagesBehind}
          </Metrics.Indicator>
        </Metrics.Section>
      </Metrics.Wrapper>
      <ControlPanelWrapper hasInput style={{ margin: '16px 0 20px' }}>
        <Search placeholder="Search by Topic Name" />
      </ControlPanelWrapper>
      <Table isFullwidth>
        <thead>
          <tr>
            <TableHeaderCell title="Topic" />
            <TableHeaderCell title="Messages behind" />
          </tr>
        </thead>
        <tbody>
          {currentPartitionsByTopic.map((key) => (
            <ListItem
              clusterName={clusterName}
              consumers={partitionsByTopic[key]}
              name={key}
              key={key}
            />
          ))}
        </tbody>
      </Table>
    </div>
  );
};

export default Details;
