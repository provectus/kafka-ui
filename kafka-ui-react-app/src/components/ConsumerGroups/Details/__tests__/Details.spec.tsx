import Details from 'components/ConsumerGroups/Details/Details';
import React from 'react';
import fetchMock from 'fetch-mock';
import { render, WithRoute } from 'lib/testHelpers';
import {
  clusterConsumerGroupDetailsPath,
  clusterConsumerGroupResetRelativePath,
} from 'lib/paths';
import { consumerGroupPayload } from 'redux/reducers/consumerGroups/__test__/fixtures';
import {
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import { act } from '@testing-library/react';

const clusterName = 'cluster1';
const { groupId } = consumerGroupPayload;

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

const renderComponent = () => {
  render(
    <WithRoute path={clusterConsumerGroupDetailsPath()}>
      <Details />
    </WithRoute>,
    { initialEntries: [clusterConsumerGroupDetailsPath(clusterName, groupId)] }
  );
};
describe('Details component', () => {
  afterEach(() => {
    fetchMock.reset();
    mockNavigate.mockClear();
  });

  describe('when consumer groups are NOT fetched', () => {
    it('renders progress bar for initial state', () => {
      fetchMock.getOnce(
        `/api/clusters/${clusterName}/consumer-groups/${groupId}`,
        404
      );
      renderComponent();
      expect(screen.getByRole('progressbar')).toBeInTheDocument();
    });
  });

  describe('when consumer gruops are fetched', () => {
    beforeEach(async () => {
      const fetchConsumerGroupMock = fetchMock.getOnce(
        `/api/clusters/${clusterName}/consumer-groups/${groupId}`,
        consumerGroupPayload
      );
      renderComponent();
      await waitForElementToBeRemoved(() => screen.getByRole('progressbar'));
      await waitFor(() => expect(fetchConsumerGroupMock.called()).toBeTruthy());
    });

    it('renders component', () => {
      expect(screen.getByRole('heading')).toBeInTheDocument();
      expect(screen.getByText(groupId)).toBeInTheDocument();

      expect(screen.getByRole('table')).toBeInTheDocument();
      expect(screen.getAllByRole('columnheader').length).toEqual(2);

      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });

    it('handles [Reset offset] click', async () => {
      userEvent.click(screen.getByText('Reset offset'));
      expect(mockNavigate).toHaveBeenLastCalledWith(
        clusterConsumerGroupResetRelativePath
      );
    });

    it('shows confirmation modal on consumer group delete', async () => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
      userEvent.click(screen.getByText('Delete consumer group'));
      await waitFor(() =>
        expect(screen.queryByRole('dialog')).toBeInTheDocument()
      );
      userEvent.click(screen.getByText('Cancel'));
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });

    it('handles [Delete consumer group] click', async () => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
      await act(() => {
        userEvent.click(screen.getByText('Delete consumer group'));
      });
      expect(screen.queryByRole('dialog')).toBeInTheDocument();
      const deleteConsumerGroupMock = fetchMock.deleteOnce(
        `/api/clusters/${clusterName}/consumer-groups/${groupId}`,
        200
      );
      await waitFor(() => {
        userEvent.click(screen.getByRole('button', { name: 'Confirm' }));
      });
      expect(deleteConsumerGroupMock.called()).toBeTruthy();

      await waitForElementToBeRemoved(() => screen.queryByRole('dialog'));
      await waitFor(() => expect(mockNavigate).toHaveBeenLastCalledWith('../'));
    });
  });
});
