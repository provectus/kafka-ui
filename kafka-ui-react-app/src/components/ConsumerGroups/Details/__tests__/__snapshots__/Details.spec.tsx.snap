// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details component when consumer gruops are NOT fetched Matches the snapshot 1`] = `
<div
  className="section"
>
  <div
    className="level"
  >
    <div
      className="level-item level-left"
    >
      <Breadcrumb
        links={
          Array [
            Object {
              "href": "/ui/clusters/local/consumer-groups",
              "label": "All Consumer Groups",
            },
          ]
        }
      >
        test
      </Breadcrumb>
    </div>
  </div>
  <PageLoader />
  <ConfirmationModal
    isOpen={false}
    onCancel={[Function]}
    onConfirm={[Function]}
  >
    Are you sure you want to delete this consumer group?
  </ConfirmationModal>
</div>
`;

exports[`Details component when consumer gruops are fetched Matches the snapshot 1`] = `
<div
  className="section"
>
  <div
    className="level"
  >
    <div
      className="level-item level-left"
    >
      <Breadcrumb
        links={
          Array [
            Object {
              "href": "/ui/clusters/local/consumer-groups",
              "label": "All Consumer Groups",
            },
          ]
        }
      >
        test
      </Breadcrumb>
    </div>
  </div>
  <div
    className="box"
  >
    <div
      className="level"
    >
      <div
        className="level-item level-right buttons"
      >
        <button
          className="button is-danger"
          onClick={[Function]}
          type="button"
        >
          Delete consumer group
        </button>
      </div>
    </div>
    <table
      className="table is-striped is-fullwidth"
    >
      <thead>
        <tr>
          <th>
            Consumer ID
          </th>
          <th>
            Host
          </th>
          <th>
            Topic
          </th>
          <th>
            Partition
          </th>
          <th>
            Messages behind
          </th>
          <th>
            Current offset
          </th>
          <th>
            End offset
          </th>
        </tr>
      </thead>
      <tbody>
        <ListItem
          clusterName="local"
          consumer={
            Object {
              "consumerId": "consumer-messages-consumer-1-122fbf98-643b-491d-8aec-c0641d2513d0",
              "currentOffset": 394,
              "endOffset": 394,
              "host": "/172.31.9.153",
              "messagesBehind": 0,
              "partition": 6,
              "topic": "messages",
            }
          }
          key="consumer-messages-consumer-1-122fbf98-643b-491d-8aec-c0641d2513d0"
        />
        <ListItem
          clusterName="local"
          consumer={
            Object {
              "consumerId": "consumer-messages-consumer-1-122fbf98-643b-491d-8aec-c0641d2513d1",
              "currentOffset": 384,
              "endOffset": 384,
              "host": "/172.31.9.153",
              "messagesBehind": 0,
              "partition": 7,
              "topic": "messages",
            }
          }
          key="consumer-messages-consumer-1-122fbf98-643b-491d-8aec-c0641d2513d1"
        />
      </tbody>
    </table>
  </div>
  <ConfirmationModal
    isOpen={false}
    onCancel={[Function]}
    onConfirm={[Function]}
  >
    Are you sure you want to delete this consumer group?
  </ConfirmationModal>
</div>
`;
